scalapb-runtime/src/main/scala/scalapb/FieldMaskTree.scala

```diff
@@ -0,0 +1,108 @@
+package scalapb
+
+import com.google.protobuf.field_mask.FieldMask
+import scalapb.FieldMaskTree.Empty
+import scalapb.descriptors.{Descriptor, PMessage, ScalaType}
+
+import scala.collection.SortedMap
+
+private[scalapb] case class FieldMaskTree(nodes: SortedMap[String, FieldMaskTree]) {
+
+  def applyToMessage[M <: GeneratedMessage: GeneratedMessageCompanion](message: M): M = {
+    val companion        = implicitly[GeneratedMessageCompanion[M]]
+    val filteredPMessage = applyToMessage(message.toPMessage)
+    companion.messageReads.read(filteredPMessage)
+  }
+
+  def containsField[M <: GeneratedMessage: GeneratedMessageCompanion](fieldNumber: Int): Boolean = {
+    val descriptor = implicitly[GeneratedMessageCompanion[M]].scalaDescriptor
+    descriptor.findFieldByNumber(fieldNumber) match {
+      case Some(field) => nodes.contains(field.name)
+      case None        => false
+    }
+  }
+
+  def fieldMask: FieldMask = FieldMask(paths)
+
+  def isValidFor[M <: GeneratedMessage: GeneratedMessageCompanion]: Boolean = {
+    val descriptor = implicitly[GeneratedMessageCompanion[M]].scalaDescriptor
+    isValidFor(descriptor)
+  }
+
+  private def applyToMessage(message: PMessage): PMessage = {
+    val fieldValues = message.value.keys.flatMap { fieldDescriptor =>
+      val value = (message.value.apply(fieldDescriptor), nodes.get(fieldDescriptor.name)) match {
+        case (pValue, Some(Empty))                 => Some(pValue)
+        case (subMessage: PMessage, Some(subTree)) => Some(subTree.applyToMessage(subMessage))
+        case _                                     => None
+      }
+      value.map(fieldDescriptor -> _)
+    }.toMap
+    PMessage(fieldValues)
+  }
+
+  private def isValidFor(descriptor: Descriptor): Boolean = {
+    nodes.forall { case (field, tree) =>
+      descriptor.findFieldByName(field) match {
+        case None => false
+        case Some(field) =>
+          if (tree == FieldMaskTree.Empty) {
+            true
+          } else {
+            field.scalaType match {
+              case ScalaType.Message(m) if !field.isRepeated => tree.isValidFor(m)
+              case _                                         => false
+            }
+          }
+      }
+    }
+  }
+
+  private def paths: Vector[String] = {
+    nodes.toVector.flatMap {
+      case (field, FieldMaskTree.Empty) => Vector(field)
+      case (field, subTree) =>
+        subTree.paths.map(path => s"$field${FieldMaskTree.FieldSeparator}$path")
+    }
+  }
+}
+
+object FieldMaskTree {
```

Comment:
should this be marked `private[scalapb]` as well, since the class is package private too?

Reply:
Yes, makes sense